[ASCII-1336] Tiny component auto exit

misteriaud opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

This PR create a wrapper component around the `autoexit` package to initialize it with fx.


### Describe how to test/QA your changes
<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->
Use the following settings in `datadog.yaml` configuration file and execute the Agent.
It should automatically stop after 30s.
```datadog.yaml
auto_exit:
  noprocess:
    enabled: true
    excluded_processes: [".*"]
  validation_period: 30
```

ogaca-dd committed:
Create agent bundle

ogaca-dd committed:
Fix lint component

ogaca-dd committed:
Add agent.Bundle() to getSharedFxOption to avoid code conflict

misteriaud committed:
creating autoexit agent component

ogaca-dd committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into olivierg/agent-bundle

misteriaud commented on comp/agent/autoexit/autoexitimpl/autoexit_test.go:
```diff
@@ -0,0 +1,90 @@
+// build servers. See #46736 for discussion.
+var fatalWaitingTime = 10 * time.Second
+
+func TestAutoexitInterrupt(t *testing.T) {
```
I created this unit test, but it takes 30s to execute, is it too long ?
It checks that the autoexit is trully sending an interupt signal.

misteriaud committed:
wrap old impl into comp

misteriaud committed:
use autoexit component in agent, process-agent, security-agent and system-probe

misteriaud committed:
fix lint issues

misteriaud committed:
Merge branch 'main' into misteriaud/ASCII-1336_tiny_component_auto_exit

misteriaud committed:
fix component lint

misteriaud committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into misteriaud/ASCII-1336_tiny_component_auto_exit

misteriaud committed:
remove useless ctx parameter in security-agent

misteriaud committed:
remove useless ctx parameter in security-agent (in windows main)

misteriaud committed:
remove useless fx.Lifecycle deps in autoexit

ogaca-dd commented on cmd/process-agent/command/main_common.go:
```diff
@@ -296,13 +301,7 @@ func initMisc(deps miscDeps) error {
 	deps.Lc.Append(fx.Hook{
-		OnStart: func(startCtx context.Context) error {
-
-			err := manager.ConfigureAutoExit(startCtx, deps.Config)
```
Did you confirm with the process agent team that using `startCtx` was a mistake?

misteriaud replied:
I asked @just-chillin, who wrote this piece of code. I didn't get her answer yet.

just-chillin replied:
I believe it is a mistake, thanks for catching it!

ogaca-dd commented on cmd/security-agent/subcommands/start/command.go:
```diff
@@ -289,12 +291,6 @@ func RunAgent(ctx context.Context, log log.Component, config config.Component, t
 		return errNoAPIKeyConfigured
 	}
 
-	err = manager.ConfigureAutoExit(ctx, config)
```
The context (`ctx`) is not the same as the one in autoexit component which is `pkgcommon.GetMainCtxCancel()`. Is that correct?

misteriaud replied:
It seems that `pkgcommon.GetMainCtxCandel()` is returning `context.WithCancel(context.Background())`

https://github.com/DataDog/datadog-agent/blob/ea5bc8bfd62cfd9f11920ea3e4f6b42369e23fe6/pkg/util/common/context.go#L26-L32

Which (if I understand well) is the same as what is provided to `RunAgent()` function in the `start()` function:

https://github.com/DataDog/datadog-agent/blob/ecc60d202478065bde3789dfad60951464c6a4e1/cmd/security-agent/subcommands/start/command.go#L214-L218

Finally the `cancel()` function is called when the `start` agent end (with defer), in `StopAgent()` function:
https://github.com/DataDog/datadog-agent/blob/ecc60d202478065bde3789dfad60951464c6a4e1/cmd/security-agent/subcommands/start/command.go#L356-L358

Am I missing something, is it different ?

ogaca-dd replied:
Can you check with the security team that this change is correct as you have one instance instead of two?

misteriaud replied:
I asked @paulcacheux and everything looks good to him.

ogaca-dd commented on comp/agent/autoexit/autoexitimpl/autoexit.go:
```diff
@@ -0,0 +1,44 @@
+}
+
+// ExitDetector is common interface for shutdown mechanisms
+type ExitDetector interface {
```
Where this interface is used?

misteriaud replied:
Thanks for noticing it ! It's a left over from a last implementation try.
I removed it in this commit: 257684cf2c3aa58f95b3aa77b1abd29739340314

brycekahle commented on cmd/system-probe/subcommands/run/command.go:
```diff
@@ -250,8 +252,7 @@ func StopSystemProbeWithDefaults() {
 	var err error
-	var ctx context.Context
-	ctx, common.MainCtxCancel = context.WithCancel(context.Background())
+	_, common.MainCtxCancel = context.WithCancel(context.Background())
```
I don't think anything is using `common.MainCtxCancel` now, so you can probably remove it.

misteriaud committed:
remove useless common.MainCtxCancel

misteriaud committed:
remove useful interface

misteriaud committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into misteriaud/ASCII-1336_tiny_component_auto_exit

misteriaud committed:
- Adding autoexit constructor to windows security agent Run function
- Remove useless cancel() context function (was used only by old autoexit impl)

misteriaud committed:
Merge branch 'main' into misteriaud/ASCII-1336_tiny_component_auto_exit

misteriaud committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into misteriaud/ASCII-1336_tiny_component_auto_exit

misteriaud commented:
/merge

misteriaud commented:
/merge
